diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index 84a3f6f..a221753 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -80,2 +80,2 @@ int ksu_handle_setresuid(uid_t ruid, uid_t euid, uid_t suid)
-        track_throne(false);
+        track_throne_sync(false);
 
diff --git a/kernel/manager/throne_tracker.c b/kernel/manager/throne_tracker.c
index 42bc844..b1d0bd0 100644
--- a/kernel/manager/throne_tracker.c
+++ b/kernel/manager/throne_tracker.c
@@ -397,19 +397,27 @@ static void ksu_throne_work_fn(struct work_struct *work)
 	}
 }
 
+bool track_throne_sync(bool prune_only)
+{
+	bool success;
+
+	mutex_lock(&throne_tracker_mutex);
+
+	const struct cred *saved_cred = override_creds(ksu_cred);
+	success = do_track_throne_core(prune_only);
+	revert_creds(saved_cred);
+
+	mutex_unlock(&throne_tracker_mutex);
+	return success;
+}
+
 void track_throne(bool prune_only)
 {
 	static bool throne_tracker_first_run __read_mostly = true;
 
 	// First scan must be synchronous to not break FDE/FBEv1 on older kernels
 	if (unlikely(throne_tracker_first_run)) {
-		mutex_lock(&throne_tracker_mutex);
-		
-		const struct cred *saved_cred = override_creds(ksu_cred);
-		do_track_throne_core(prune_only);
-		revert_creds(saved_cred);
-		
-		mutex_unlock(&throne_tracker_mutex);
+		track_throne_sync(prune_only);
 		throne_tracker_first_run = false;
 		return;
 	}
diff --git a/kernel/manager/throne_tracker.h b/kernel/manager/throne_tracker.h
index c937317..6394b0c 100644
--- a/kernel/manager/throne_tracker.h
+++ b/kernel/manager/throne_tracker.h
@@ -14,11 +14,17 @@ static inline void track_throne(bool prune_only)
     (void)prune_only;
 }
+static inline bool track_throne_sync(bool prune_only)
+{
+    (void)prune_only;
+    return false;
+}
 #else
 void ksu_throne_tracker_init();
 
 void ksu_throne_tracker_exit();
 
 void track_throne(bool prune_only);
+bool track_throne_sync(bool prune_only);
 #endif
 
 #endif
