diff --git a/kernel/feature/sucompat.c b/kernel/feature/sucompat.c
--- a/kernel/feature/sucompat.c
+++ b/kernel/feature/sucompat.c
@@ -107,5 +107,4 @@ int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
 	if (unlikely(!memcmp(path, su, sizeof(su)))) {
 		ksu_compat_sulog('a');
-		pr_info("faccessat su->sh!\n");
 		*filename_user = sh_user_path();
 	}
@@ -137,5 +136,4 @@ int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags)
 	if (unlikely(!memcmp(path, su, sizeof(su)))) {
 		ksu_compat_sulog('s');
-		pr_info("newfstatat su->sh!\n");
 		*filename_user = sh_user_path();
 	}
@@ -234,7 +232,6 @@ long ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
 	if (likely(memcmp(filename->name, su, sizeof(su))))
 		return 0;
 
-	pr_info("do_execveat_common su found\n");
 	memcpy((void *)filename->name, ksud_path, sizeof(ksud_path));
 
 	escape_with_root_profile();
